File: 70.climbing-stairs.js
/*
 * @lc app=leetcode id=70 lang=javascript
 *
 * [70] Climbing Stairs
 */

// @lc code=start
/**
 * @param {number} n
 * @return {number}
 */
var climbStairs = function(n) {
  let cache = []
  function climb(n) {
    if (cache[n]) return cache[n]
    if (n<3) return n
    cache[n] = climb(n-2) + climb(n-1)
    return cache[n]
  }

  return climb(n)
};
// @lc code=end

File: 70.climbing-stairs.py
#
# @lc app=leetcode id=70 lang=python3
#
# [70] Climbing Stairs
#

# @lc code=start
class Solution:
    def __init__(self):
        self.ns = {}

    def climbStairs(self, n: int) -> int:
        if n < 0: return 0
        elif n <= 1: return 1

        if n not in self.ns:
            self.ns[n] = self.climbStairs(n-2) + self.climbStairs(n-1)
        return self.ns[n]
# @lc code=end

